### ISSUE TITLE
The pointer passed to super type hook is not converted to the super type

### ISSUE THREAD

**Issue by alaviss** [nim-works/nimskull#1337]
GCC 14 enforces a new language change that prohibit implicit conversion between incompatible pointer types in C. Some codegen that used to work no longer does.

### Example

```nim
type X = object of RootObj
proc `=destroy`(x: var X) = discard

type Y = object of X
let y = Y()
```

### Actual Output

```
@mtest.nim.c: In function ‘eqdestroy___test_6’:
@mtest.nim.c:127:28: error: passing argument 1 of ‘eqdestroy___test_2’ from incompatible pointer type [-Wincompatible-pointer-types]
  127 |         eqdestroy___test_2(dest);
      |                            ^~~~
      |                            |
      |                            tyObject_Y__mnLzXY1lPCpgpET0c8sT3w *
@mtest.nim.c:62:87: note: expected ‘tyObject_X__oKPdmARtpzUKjBTlF51xpw *’ but argument is of type ‘tyObject_Y__mnLzXY1lPCpgpET0c8sT3w *’
   62 | N_LIB_PRIVATE N_NIMCALL(void, eqdestroy___test_2)(tyObject_X__oKPdmARtpzUKjBTlF51xpw* x);
      |                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
@mtest.nim.c: In function ‘eqtrace___test_15’:
@mtest.nim.c:130:27: error: passing argument 1 of ‘eqtrace___test_27’ from incompatible pointer type [-Wincompatible-pointer-types]
  130 |         eqtrace___test_27(dest, env);
      |                           ^~~~
      |                           |
      |                           tyObject_Y__mnLzXY1lPCpgpET0c8sT3w *
@mtest.nim.c:63:86: note: expected ‘tyObject_X__oKPdmARtpzUKjBTlF51xpw *’ but argument is of type ‘tyObject_Y__mnLzXY1lPCpgpET0c8sT3w *’
   63 | N_LIB_PRIVATE N_NIMCALL(void, eqtrace___test_27)(tyObject_X__oKPdmARtpzUKjBTlF51xpw* dest, void* env);
      |                                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
```

### Possible Solution

- Cast from Y to X before calling the super type hook
- Disable the new warnings in `nimbase.h`

### References
- [Porting to GCC 14#Type checking on pointer types](https://gcc.gnu.org/gcc-14/porting_to.html#incompatible-pointer-types)

**Comment by mrgaturus** [comment]
notice clang prohibited implicit conversion many years ago, so nimskull used a gcc-ism all this time

**Comment by zerbina** [comment]
I'm on it.

It's a relatively minor thing, but the bug is with hook synthesis (`liftdestructors`), which is currently not part of the backend, so I adjusted the label.

**Comment by alaviss** [comment]
Fixed by #1338
